### PR TITLE
Reduce duplicate links by allowing pending links a short window to be…

### PR DESCRIPTION
**Pull request by plorenz**
… completed before we try again

### CODE REVIEW

File: controller/network/link_controller.go [hunk 1]
```diff
@@ -141,28 +148,24 @@ func (linkController *linkController) missingLinks(routers []*Router) ([]*Link,
 	return missingLinks, nil
 }
 
-func (linkController *linkController) firstDirectedLink(a, b *Router) (*Link, bool) {
-	// a->b
+func (linkController *linkController) hasLink(a, b *Router, pendingLimit int64) bool {
+	return linkController.hasDirectedLink(a, b, pendingLimit) || linkController.hasDirectedLink(b, a, pendingLimit)
```
Comment (andrewpmartinez) [inline controller/network/link_controller.go:152]: What a subtle difference in the clauses.

**Comment by andrewpmartinez** [comment]
I don't know if there are any side effects of this code. My approval is so we can get this into soak testing.

**Review by andrewpmartinez — APPROVED** [review]

File: controller/network/link_controller.go [hunk 2]
```diff
@@ -120,11 +121,17 @@ func (linkController *linkController) leastExpensiveLink(a, b *Router) (*Link, b
 }
 
 func (linkController *linkController) missingLinks(routers []*Router) ([]*Link, error) {
+	// When there's a flood of router connects as startup we can see the same link
+	// as missing multiple times as the new link will be marked as PENDING until it's
+	// connected. Give ourselves a little window to make the connection before we
+	// send another dial
+	pendingLimit := info.NowInMilliseconds() - 5000
```
[the remark this answers is not in the document]
Comment (plorenz) [reply controller/network/link_controller.go]: Added a config setting for pending link timeout

File: controller/network/link_controller.go [hunk 3]
```diff
@@ -119,12 +121,18 @@ func (linkController *linkController) leastExpensiveLink(a, b *Router) (*Link, b
 	return nil, false
 }
 
-func (linkController *linkController) missingLinks(routers []*Router) ([]*Link, error) {
+func (linkController *linkController) missingLinks(routers []*Router, pendingTimeout time.Duration) ([]*Link, error) {
+	// When there's a flood of router connects as startup we can see the same link
```
Comment (michaelquigley) [inline controller/network/link_controller.go:125]: at startup

**Review by michaelquigley — APPROVED** [review]

File: controller/network/link_controller.go (hunk 3 above)
Comment (plorenz) [reply controller/network/link_controller.go]: fixed

**Pull request merged**